src/hotspot/share/cds/archiveBuilder.cpp

```diff
@@ -1052,39 +1051,38 @@ class ArchiveBuilder::CDSMapLogger : AllStatic {
-
-      while (start < end) {
-        size_t byte_size;
-        oop original_oop = ArchiveHeapWriter::buffered_addr_to_source_obj(start);
-        if (original_oop != nullptr) {
-          ResourceMark rm;
-          log_info(cds, map)(PTR_FORMAT ": @@ Object %s",
-                             p2i(to_requested(start)), original_oop->klass()->external_name());
-          byte_size = original_oop->size() * BytesPerWord;
-        } else if (start == ArchiveHeapWriter::buffered_heap_roots_addr()) {
-          // HeapShared::roots() is copied specially so it doesn't exist in
-          // HeapShared::OriginalObjectTable. See HeapShared::copy_roots().
-          log_info(cds, map)(PTR_FORMAT ": @@ Object HeapShared::roots (ObjArray)",
-                             p2i(to_requested(start)));
-          byte_size = ArchiveHeapWriter::heap_roots_word_size() * BytesPerWord;
-        } else {
-          // We have reached the end of the region
-          break;
-        }
-        address oop_end = start + byte_size;
-        log_data(start, oop_end, to_requested(start), /*is_heap=*/true);
-        start = oop_end;
-      }
-      if (start < end) {
-        log_info(cds, map)(PTR_FORMAT ": @@ Unused heap space " SIZE_FORMAT " bytes",
-                           p2i(to_requested(start)), size_t(end - start));
-        log_data(start, end, to_requested(start), /*is_heap=*/true);
+  static void log_heap_region(ArchiveHeapInfo* heap_info) {
+    MemRegion r = heap_info->memregion();
+    address start = address(r.start());
+    address end = address(r.end());
+    log_region("heap", start, end, to_requested(start));
+
+    while (start < end) {
+      size_t byte_size;
+      oop original_oop = ArchiveHeapWriter::buffered_addr_to_source_obj(start);
+      if (original_oop != nullptr) {
+        ResourceMark rm;
+        log_info(cds, map)(PTR_FORMAT ": @@ Object %s",
+                           p2i(to_requested(start)), original_oop->klass()->external_name());
+        byte_size = original_oop->size() * BytesPerWord;
+      } else if (start == ArchiveHeapWriter::buffered_heap_roots_addr()) {
+        // HeapShared::roots() is copied specially so it doesn't exist in
+        // HeapShared::OriginalObjectTable. See HeapShared::copy_roots().
+        log_info(cds, map)(PTR_FORMAT ": @@ Object HeapShared::roots (ObjArray)",
+                           p2i(to_requested(start)));
+        byte_size = ArchiveHeapWriter::heap_roots_word_size() * BytesPerWord;
+      } else {
+        // We have reached the end of the region
+        break;
       }
+      address oop_end = start + byte_size;
+      log_data(start, oop_end, to_requested(start), /*is_heap=*/true);
+      start = oop_end;
+    }
+    if (start < end) {
+      log_info(cds, map)(PTR_FORMAT ": @@ Unused heap space " SIZE_FORMAT " bytes",
+                         p2i(to_requested(start)), size_t(end - start));
+      log_data(start, end, to_requested(start), /*is_heap=*/true);
     }
```

Comment:
These log messages can be placed inside the else case before the break

Reply:
Fixed.